src/bream/core.py

```diff
@@ -0,0 +1,375 @@
+
+    type_label: TypeLabel
+    """A label for the type `T` this codec handles.
+
+    This should be an identifier that is unique within a format. It can NEVER be changed
+    once specified if wanting to maintain backwards compatibility for this format.
+    """
+
+    type_spec: TypeSpec
+    """The `TypeSpec` for the type `T` that this codec handles.
+
+    Note that changing the `TypeSpec` need not trigger a bump in the `version`. For
+    example, an `ImportableType` might be renamed, or moved to another module. This
+    means that the type spec changes.
+    """
+
+    # FIXME: generic on Coder might need to be covariant? We might re-use the same
+    #   coder for multiple concrete coders.
+    coder: Coder[T]
+    """A strategy for encoding and decoding instances of `T`."""
+
+
+@typing.final
+class SerialisationFormat:
+    """A serialisation format is conceptually a collection of coders.
+
+    No two coders should have the same label, and no two coders should operate on the
+    same type.
+    """
+
+    def __init__(self, *, codecs: Iterable[Codec[Any]]) -> None:
+        spec_to_codec: dict[TypeSpec, Codec[Any]] = {}
+        label_to_codec: dict[TypeLabel, Codec[Any]] = {}
+        for codec in codecs:
+            spec = codec.type_spec
+            if _is_native_type(spec):
+                msg = f"cannot add codec for native type: {spec}"
+                raise ValueError(msg)
+            label = codec.type_label
+            if spec in spec_to_codec:
+                msg = f"multiple codecs for type spec: {spec}"
+                raise ValueError(msg)
+            if label in label_to_codec:
+                msg = f"multiple codecs for type label: {label}"
+                raise ValueError(msg)
+            spec_to_codec[spec] = codec
+            label_to_codec[label] = codec
+        self._spec_to_codec = spec_to_codec
+        self._label_to_codec = label_to_codec
+
+    def find_codec_for_value[T](self, obj: T) -> Codec[T] | None:
+        """Find a suitable coder for `obj`, or `None` if there isn't one."""
+        spec = TypeSpec.from_type(type(obj))
+        return self._spec_to_codec.get(spec)
+
+    def find_codec_for_type_label(self, type_label: TypeLabel) -> Codec[Any] | None:
+        """Find a suitable coder for `type_label`, or `None` if there isn't one."""
+        return self._label_to_codec.get(type_label)
+
+
+def encode_to_document(obj: object, fmt: SerialisationFormat) -> EncodeError | Document:
+    """Encode `obj`, and place inside an bream document."""
+    payload = encode(obj, fmt)
+    if isinstance(payload, EncodeError):
+        return payload
+
+    return {Keys.bream_spec.value: BREAM_SPEC, Keys.payload.value: payload}
+
+
+def encode(obj: object, fmt: SerialisationFormat) -> EncodeError | JsonType:
+    # FIXME: isinstance is too lax here; should use type identity
+    if isinstance(obj, _JsonElement):
+        return obj
+
+    # FIXME: isinstance is too lax here; should use type identity
```

Comment:
This FIXME, and the one above, is new.

The point being is that if a user creates a type that inherits from `list` then they must be forced to provide their own coder to serialise this. We must not go via this pathway.

Reply:
I see several other isinstance checks that might have this failure mode

Reply:
#17